src/systems/systems/src/pressurization/cabin_pressure_controller.rs

```diff
@@ -0,0 +1,455 @@
+
+        let rate_of_change_for_delta = INTERNAL_VS_RATE_CHANGE * context.delta_as_secs_f64();
+        if self.cabin_target_vs > self.cabin_vs() {
+            self.cabin_vs
+                + Velocity::new::<foot_per_minute>(rate_of_change_for_delta.min(
+                    self.cabin_target_vs.get::<foot_per_minute>()
+                        - self.cabin_vs().get::<foot_per_minute>(),
+                ))
+        } else if self.cabin_target_vs < self.cabin_vs() {
+            self.cabin_vs
+                - Velocity::new::<foot_per_minute>(rate_of_change_for_delta.min(
+                    self.cabin_vs().get::<foot_per_minute>()
+                        - self.cabin_target_vs.get::<foot_per_minute>(),
+                ))
+        } else {
+            self.cabin_target_vs
+        }
+    }
+
+    pub fn cabin_pressure(&self) -> Pressure {
+        self.cabin_pressure
+    }
+
+    pub fn cabin_altitude(&self) -> Length {
+        self.cabin_alt
+    }
+
+    pub fn cabin_vs(&self) -> Velocity {
+        self.cabin_vs
+    }
+
+    pub fn cabin_delta_p(&self) -> Pressure {
+        self.cabin_pressure - self.exterior_pressure
+    }
+
+    pub fn pressure_schedule(&self) -> PressureSchedule {
+        self.pressure_schedule_manager.pressure_schedule()
+    }
+
+    pub fn should_switch_cpc(&self) -> bool {
+        self.pressure_schedule_manager.should_switch_cpc()
+    }
+
+    pub fn reset_cpc_switch(&mut self) {
+        self.pressure_schedule_manager.reset_cpc_switch();
+    }
+}
+
+impl PressureValveActuator for CabinPressureController {
+    fn should_open_pressure_valve(&self) -> bool {
+        self.cabin_vs < Velocity::new::<meter_per_second>(0.)
+            || self.pressure_schedule() == PressureSchedule::Ground
+    }
+
+    fn should_close_pressure_valve(&self) -> bool {
+        self.cabin_vs > Velocity::new::<meter_per_second>(0.)
+    }
+
+    fn target_valve_position(&self) -> Ratio {
+        // Placeholder formula to simulate OFV behaviour based on empirical data
+        let pressure_ratio: f64 = self.exterior_pressure.get::<hectopascal>()
+            / (self.cabin_pressure().get::<hectopascal>() * 2.);
+        let vs_ratio: f64 = self.cabin_target_vs.get::<foot_per_minute>() / 2000. * pressure_ratio;
+        if (pressure_ratio + vs_ratio) > 1.
+            || (self.pressure_schedule() == PressureSchedule::Ground
+                && self.pressure_schedule_manager.should_open_outflow_valve())
+        {
+            Ratio::new::<percent>(100.)
+        } else if (pressure_ratio + vs_ratio) < 0. {
+            Ratio::new::<percent>(0.)
+        } else {
+            Ratio::new::<percent>((pressure_ratio + vs_ratio) * 100.)
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PressureSchedule {
+    Ground,
+    TakeOff,
+    ClimbInternal,
+    Cruise,
+    DescentInternal,
+    Abort,
+}
+
+#[derive(Copy, Clone)]
+struct PressureScheduleManager {
```

Comment:
I'm still thinking that my earlier comment on a state machine is valid. Via Discord you asked:
> so for example the function should_open_pressure_valve(..) needs to know if the aircraft is on the ground, and I can't just call self.pressure_schedule() == PressureSchedule::Ground anymore.

You can (considering an enum state machine) write:

```rust
impl PressureSchedule {
    fn is_ground(&self) -> bool {
        matches!(self, PressureSchedule::Ground)
    }
}
```

Reply:
After considerable reading and learning 😄 I've managed to implement some sort of state machine for the `PressureSchedule` within the CPC. I've tested functionality and it all seems to work. I understand this is a better solution to control state transitions, but the amount of code has increased significantly! Feel free to reach out on Discord if you want to discuss this in more detail.

**Note**: I've left a "shared variable" called `vertical_speed` within the struct. At the moment it's unused but I plan to submit a new refactor within the coming days to combine this new `PressureSchedule` with the vertical speed calculation.